### DISCUSSION TITLE
NeighborLoader - k-hop neighbors

### DISCUSSION THREAD

**Discussion by Vid201** [pyg-team/pytorch_geometric#3249] [Q&A]
Hi, is it possible to construct bipartite graph (like returned from NeighborSampler) with batch from NeighborLoader? How to know to which depth sampled nodes/edges belong?

**Comment by rusty1s** [comment]
Currently, there is no such way, as the `NeighborLoader` tries to simplify the GNN model implementation in a neighbor sampling scenario. Can you explain a bit more what's the benefit of such information?

One thing I could imagine is that we add a `sizes` attribute to the returned data (holding information about which slices of nodes where sampled in which iteration), but this requires both changes to PyG and `torch-sparse`.

**Reply by Vid201** [reply]
Ok, it seems reasonable, but when doing experiments on the large-scale dataset and multi-layer GNN, the model's accuracy is much lower when doing message passing over whole heterogeneous edge index returned from `NeighborLoader`, than doing message passing over heterogeneous bipartite graph returned from `NeighborSampler` (custom implementation to support heterogeneous graphs based on `NeighborSampler`'s implementation). So in my case, it is crucial to distinguish nodes/edges from different depths.

**Reply by rusty1s** [reply]
The model outputs of `NeighborLoader` and a heterogeneous `NeighborSampler` variant (within bipartite graphs) should be identical (given that we sample the same neighbors in each iteration). I'm curious why there should be a drop in performance using `NeighborLoader`. Can you share your `NeighborSampler` code with me?

**Reply by Vid201** [reply]
Sure, I can share. What exactly do you mean within bipartite graphs? Because I am not doing sampling on bipartite graphs, I just need the bipartite computation graph, like the one returned from `NeighborSampler`.

Link to the code: https://codeshare.io/6pmYJ0 (I can provide more comments if necessary - some variables are badly named, code was only tested with `SparseTensor` as input and output).

The general idea is to sample per each relation/edge type and then merge multiple `adj_t` on each depth separately (relation type is saved in value for `SparseTensor`), ending with bipartite computation graph (same as `NeighborSampler`).

**Reply by rusty1s** [reply]
Thanks for sharing. Do you also have some code snippets that show-cases performance differences between your heterogeneous `NeighborSampler` and our heterogeneous `NeighborLoader`? To me, this sounds like they are doing the exact same thing, so I am keen to investigate your observed performance differences.

**Reply by Vid201** [reply]
Yes, I have. Here is the example for MAG240M from OGB-LSC (code structure is the same as the example on OGB GitHub): https://drive.google.com/file/d/1yuRojLfMP_ceNL2Kbi2Pae6cSl2SGkyj/view?usp=sharing . You should be able to open the code with Google Colab.

Thanks a lot for looking into it! The difference in validation accuracy is 4.6 % (lower with NeighborLoader) after 2 epochs. I am doing the batch conversion so the same model can be reused for both neighbor loaders.

I think the difference is if node A is sampled in iteration 1 and then again in iteration 2, with `NeighborLoader` it will aggregate messages from 2 times the defined neighborhood size on each layer (since only one adj. matrix is returned). While with `NeighborSampler`, it will aggregate right amount of messages on each layer (and each time from different neighbors). If I understand the implementation correctly.

I hope I do not have some unexpected bug in the code :o

**Reply by rusty1s** [reply]
Thanks for pointing me to the scripts. This helps me tremendously. I will dig into it and let you know.

**Reply by ilyalasy** [reply]
Hi, @rusty1s, is _iterations_ in NeighborLoader documentation equals to _hops_? If I pass _num_neighbors_ = [-1,-1], does it mean that it will take all neighbors of a node and all neighbors of neighbors?

**Reply by rusty1s** [reply]
Yes, that is correct.

**Reply by MDrance** [reply]
Hello,
I'm sorry to answer an ancient topic but I have a question related to Misterion777 question about the number of iterations in the _NeighborLoader_. Is the number or iterations or hops something I can have an impact on directly using the _NeighborLoader_ class, or is it defined by default by the number of layers in my model ?
I'm currently working with a simple _GAE_, I have embeddings for my nodes that I'm trying to encode and decode, planing on using the output of my encoder for an other task. I would like to take the information about nodes in 2 iterations/hops away from current node and I'm not sure how to do it.
If I understand correctly the computation graph for a node and the number of layers, the code below should use the neighbors of the current node and their neighbors too ? 

```
class EncoderGATConv(torch.nn.Module):
    def __init__(self, in_channels, out_channels):
        super(EncoderGATConv, self).__init__()
        self.conv1 = GATConv(in_channels, 2 * out_channels)
        self.conv2 = GATConv(2 * out_channels, out_channels)
    
    def forward(self, data):        
        x = data.x
        edge_index = data.edge_index
        z = self.conv1(x, edge_index).relu()
        z = self.conv2(x, edge_index)    
        return z
    
class DecoderGATConv(torch.nn.Module):        
    def __init__(self, in_channels, out_channels):
        super(DecoderGATConv, self).__init__()        
        self.conv1 = GATConv(in_channels, 2 * out_channels)
        self.conv2 = GATConv(2 *  out_channels, out_channels)
        
    def forward(self, z, edge_index):        
        out = self.conv1(z, edge_index).relu()
        out = self.conv2(out, edge_index)
        return out
```
Best,
Martin

**Reply by Padarn** [reply]
This is defined in your `NeighborLoader`:
```
        num_neighbors (List[int] or Dict[Tuple[str, str, str], List[int]]): The
            number of neighbors to sample for each node in each iteration.
            In heterogeneous graphs, may also take in a dictionary denoting
            the amount of neighbors to sample for each individual edge type.
            If an entry is set to :obj:`-1`, all neighbors will be included.
````

**Reply by MDrance** [reply]
Hello Padarn and thanks for your answer,
I understand that the _NeighborLoader_ allow me to chose the number of neighbors to sample at each iterations/hops, but I'm not sure about the number of iterations/hops I'm using right now ? 
In the computation graph in example below, to update yellow node features, I want to take the direct neighbors features (blue nodes) and the green node features too (so 2 iterations/hops from the current node).

![image](https://user-images.githubusercontent.com/57141302/190972708-84fbb952-84e9-4045-9275-df37a7097860.png)

**Reply by Padarn** [reply]
The number of neighbours your model uses will depend on your graph. So I guess it depends on the maximum degree of a node in your data.
